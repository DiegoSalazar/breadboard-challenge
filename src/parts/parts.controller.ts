import {
  Controller,
  Get,
  Query,
  HttpException,
  HttpStatus,
} from '@nestjs/common';

import { PartsService } from './parts.service';

@Controller('parts')
export class PartsController {
  constructor(private readonly partsService: PartsService) {}

  @Get('aggregate')
  async getAggregatedParts(@Query('partNumber') partNumber: string) {
    if (!partNumber) {
      throw new HttpException(
        'Part number is required',
        HttpStatus.BAD_REQUEST,
      );
    }

    try {
      return await this.partsService.fetchAndAggregate(partNumber);
    } catch (error) {
      throw new HttpException(error.message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}
