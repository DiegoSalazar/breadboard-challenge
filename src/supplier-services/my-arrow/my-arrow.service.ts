import { Injectable } from '@nestjs/common'
import { HttpService } from '@nestjs/axios'
import { firstValueFrom } from 'rxjs'

import { AggregatedPart } from '../../interfaces/aggregated-part.interface'
import { MyArrowProduct } from '../../interfaces/my-arrow-part.interface'

type MyArrowResponse = {
  status: string
  pricingResponse: MyArrowProduct[]
}

@Injectable()
export class MyArrowService {
  private readonly url =
    'https://backend-takehome.s3.us-east-1.amazonaws.com/myarrow.json'

  constructor(private readonly httpService: HttpService) {}

  async findParts(partNumber: string): Promise<AggregatedPart[]> {
    const { data } = await firstValueFrom(
      this.httpService.get<MyArrowResponse>(this.url),
    );

    if (data.status !== 'SUCCESS') {
      throw Error('Failed to fetch parts data from MyArrow');
    }

    const parts = data.pricingResponse.filter(
      (p) => p.partNumber === partNumber,
    );

    return parts.map(p => {
      const aggregatedPart: AggregatedPart = {
        name: p.partNumber,
        description: p.description,
        totalStock: Number(p.fohQuantity),
        manufacturerLeadTime: p.leadTime?.supplierLeadTime,
        manufacturerName: p.manufacturer,
        packaging: p.pricingTier?.map((tier) => ({
          supplier: 'Arrow',
          type: p.pkg,
          minimumOrderQuantity: tier.minQuantity,
          quantityAvailable: tier.maxQuantity,
          unitPrice: tier.resalePrice,
          priceBreaks: [],
        })),
        productDoc:
          p.urlData.find((d) => d.type.toLowerCase() === 'datasheet')?.value ??
          '',
        productUrl:
          p.urlData.find((d) => d.type.toLowerCase() === 'part details')
            ?.value ?? '',
        productImageUrl: '',
        // @ts-ignore TODO: which fields are the specs?
        specifications: [],
        sourceParts: [],
      };
      return aggregatedPart;
    })
  }
}
