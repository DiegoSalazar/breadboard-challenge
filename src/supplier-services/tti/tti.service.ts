import { Injectable } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';

import { TtiPart } from '../../interfaces/tti-part.interface';
import { AggregatedPart } from '../../interfaces/aggregated-part.interface';

type TtiResponse = { parts: TtiPart[] };

@Injectable()
export class TtiService {
  private readonly url =
    'https://backend-takehome.s3.us-east-1.amazonaws.com/tti.json';

  constructor(private readonly httpService: HttpService) {}

  async findParts(partNumber: string): Promise<AggregatedPart[]> {
    const { data } = await firstValueFrom(
      this.httpService.get<TtiResponse>(this.url),
    );
    const parts = data.parts.filter((p) =>
      [p.ttiPartNumber, p.manufacturerPartNumber].includes(partNumber),
    );

    return parts.map((p) => {
      const aggregatedPart: AggregatedPart = {
        name: p.manufacturerPartNumber,
        description: p.description,
        totalStock: p.availableToSell,
        manufacturerLeadTime: Number(p.leadTime),
        manufacturerName: p.manufacturer,
        packaging: [
          {
            supplier: 'TTI',
            type: p.packaging,
            minimumOrderQuantity: p.salesMinimum,
            quantityAvailable: p.availableToSell,
            unitPrice: Number(p.pricing.vipPrice),
            priceBreaks: p.pricing.quantityPriceBreaks,
          },
        ],
        productDoc: p.datasheetURL,
        productUrl: p.buyUrl,
        productImageUrl: p.imageURL,
        // @ts-ignore TODO: which fields are the specs?
        specifications: p.datasheetURL,
        sourceParts: [],
      };
      return aggregatedPart;
    });
  }
}
