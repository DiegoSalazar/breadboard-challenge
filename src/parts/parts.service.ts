import { Injectable } from '@nestjs/common';

import { TtiService } from '../supplier-services/tti/tti.service';
import { MyArrowService } from '../supplier-services/my-arrow/my-arrow.service';
import { AggregatedPart } from '../interfaces/aggregated-part.interface';

type SupplierService = (TtiService | MyArrowService)[] ;

@Injectable()
export class PartsService {
  private readonly suppliers: SupplierService;

  constructor(
    private readonly ttiService: TtiService,
    private readonly myArrowService: MyArrowService,
  ) {
    // quick and dirty registration of supplier services
    // to scale up, should iterate over a directory of services of the same Supplier superclass
    this.suppliers = [ttiService, myArrowService];
  }

  async fetchAndAggregate(partNumber: string) {
    const results = await Promise.allSettled(
      this.suppliers.map((service) => service.findParts(partNumber)),
    );
    const parts = results
      .reduce((agg, r) => [...(r.status === 'fulfilled' ? r.value : []), ...agg], []);

    if (parts.length === 0) {
      throw new Error(`No data found for part number ${partNumber}`);
    }

    const aggregatedPart: AggregatedPart = {
      name: parts[0].name,
      description: parts[0].description,
      totalStock: parts.reduce((agg, p) => agg + p.totalStock, 0),
      manufacturerLeadTime: Math.min(...parts.map((p) => p.manufacturerLeadTime)),
      manufacturerName: parts[0].manufacturerName,
      packaging: parts.reduce((agg, p) => [...agg, ...p.packaging ?? []], []),
      productDoc: parts[0].productDoc,
      productUrl: parts[0].productUrl,
      productImageUrl: parts[0].productImageUrl,
      specifications: parts[0].specifications,
      sourceParts: parts[0].sourceParts,
    }

    return aggregatedPart;
  }
}
