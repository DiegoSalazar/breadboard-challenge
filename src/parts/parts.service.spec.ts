import { Test, TestingModule } from '@nestjs/testing';
import { HttpModule } from '@nestjs/axios';
import { PartsService } from './parts.service';
import { TtiService } from '../supplier-services/tti/tti.service';
import { MyArrowService } from '../supplier-services/my-arrow/my-arrow.service';

describe('PartsService', () => {
  let service: PartsService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [HttpModule],
      providers: [PartsService, TtiService, MyArrowService],
    }).compile();

    service = module.get<PartsService>(PartsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('returns parts of given part number', async () => {
    const partNumber = '0510210200';
    const parts = await service.fetchAndAggregate(partNumber);
    expect(parts).toMatchSnapshot();
  });
});
