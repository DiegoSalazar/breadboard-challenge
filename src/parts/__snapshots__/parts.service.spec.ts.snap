// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartsService returns parts of given part number 1`] = `
[
  {
    "description": "Conn Housing RCP 2 POS 1.25mm Crimp ST Cable Mount Natural PicoBlade™ Bag",
    "manufacturerLeadTime": 8,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": [
      {
        "minimumOrderQuantity": "1",
        "priceBreaks": [],
        "quantityAvailable": "856",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0773",
      },
      {
        "minimumOrderQuantity": "857",
        "priceBreaks": [],
        "quantityAvailable": "4467",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0725",
      },
      {
        "minimumOrderQuantity": "4468",
        "priceBreaks": [],
        "quantityAvailable": "15220",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0651",
      },
      {
        "minimumOrderQuantity": "15221",
        "priceBreaks": [],
        "quantityAvailable": "24488659620",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0545",
      },
    ],
    "productDoc": "https://download.siliconexpert.com/pdfs2/2023/3/7/6/25/48/761619/mol_/manual/0510210200_crimp_housings.pdf",
    "productImageUrl": "",
    "productUrl": "https://my.arrow.com/products/2384_06516206",
    "sourceParts": [],
    "specifications": [],
    "totalStock": 0,
  },
  {
    "description": "Conn Housing RCP 2 POS 1.25mm Crimp ST Cable Mount Natural PicoBlade™ Bag",
    "manufacturerLeadTime": 8,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": [
      {
        "minimumOrderQuantity": "1",
        "priceBreaks": [],
        "quantityAvailable": "856",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0773",
      },
      {
        "minimumOrderQuantity": "857",
        "priceBreaks": [],
        "quantityAvailable": "4467",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0725",
      },
      {
        "minimumOrderQuantity": "4468",
        "priceBreaks": [],
        "quantityAvailable": "15220",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0651",
      },
      {
        "minimumOrderQuantity": "15221",
        "priceBreaks": [],
        "quantityAvailable": "24488659620",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0545",
      },
    ],
    "productDoc": "https://download.siliconexpert.com/pdfs2/2023/3/7/6/25/48/761619/mol_/manual/0510210200_crimp_housings.pdf",
    "productImageUrl": "",
    "productUrl": "https://my.arrow.com/products/2381_06516206",
    "sourceParts": [],
    "specifications": [],
    "totalStock": 0,
  },
  {
    "description": "Conn Housing RCP 2 POS 1.25mm Crimp ST Cable Mount Natural PicoBlade™ Bag",
    "manufacturerLeadTime": 10,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": [
      {
        "minimumOrderQuantity": "1",
        "priceBreaks": [],
        "quantityAvailable": "579",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.114",
      },
      {
        "minimumOrderQuantity": "580",
        "priceBreaks": [],
        "quantityAvailable": "3026",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.107",
      },
      {
        "minimumOrderQuantity": "3027",
        "priceBreaks": [],
        "quantityAvailable": "4999",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0961",
      },
      {
        "minimumOrderQuantity": "5000",
        "priceBreaks": [],
        "quantityAvailable": "10999",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0828",
      },
      {
        "minimumOrderQuantity": "11000",
        "priceBreaks": [],
        "quantityAvailable": "12457",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0795",
      },
      {
        "minimumOrderQuantity": "12458",
        "priceBreaks": [],
        "quantityAvailable": "21999",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0666",
      },
      {
        "minimumOrderQuantity": "22000",
        "priceBreaks": [],
        "quantityAvailable": "199999",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0638",
      },
      {
        "minimumOrderQuantity": "200000",
        "priceBreaks": [],
        "quantityAvailable": "999999999",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.061",
      },
    ],
    "productDoc": "https://download.siliconexpert.com/pdfs2/2023/3/7/6/25/48/761619/mol_/manual/0510210200_crimp_housings.pdf",
    "productImageUrl": "",
    "productUrl": "https://my.arrow.com/products/1790_18300573",
    "sourceParts": [],
    "specifications": [],
    "totalStock": 0,
  },
  {
    "description": "Conn Housing RCP 2 POS 1.25mm Crimp ST Cable Mount Natural PicoBlade™ Bag",
    "manufacturerLeadTime": undefined,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": undefined,
    "productDoc": "https://download.siliconexpert.com/pdfs2/2023/3/7/6/25/48/761619/mol_/manual/0510210200_crimp_housings.pdf",
    "productImageUrl": "",
    "productUrl": "https://my.arrow.com/products/2348_06516206",
    "sourceParts": [],
    "specifications": [],
    "totalStock": 0,
  },
  {
    "description": "Conn Housing RCP 2 POS 1.25mm Crimp ST Cable Mount Natural PicoBlade™ Bag",
    "manufacturerLeadTime": 8,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": [
      {
        "minimumOrderQuantity": "1",
        "priceBreaks": [],
        "quantityAvailable": "856",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0773",
      },
      {
        "minimumOrderQuantity": "857",
        "priceBreaks": [],
        "quantityAvailable": "4467",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0725",
      },
      {
        "minimumOrderQuantity": "4468",
        "priceBreaks": [],
        "quantityAvailable": "15220",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0651",
      },
      {
        "minimumOrderQuantity": "15221",
        "priceBreaks": [],
        "quantityAvailable": "24488659620",
        "supplier": "Arrow",
        "type": "EACH",
        "unitPrice": "0.0545",
      },
    ],
    "productDoc": "https://download.siliconexpert.com/pdfs2/2023/3/7/6/25/48/761619/mol_/manual/0510210200_crimp_housings.pdf",
    "productImageUrl": "",
    "productUrl": "https://my.arrow.com/products/1790_06516206",
    "sourceParts": [],
    "specifications": [],
    "totalStock": 385887,
  },
  {
    "description": "Headers & Wire Housings 1.25MM REC 02P FEMALE",
    "manufacturerLeadTime": NaN,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": [
      {
        "minimumOrderQuantity": 42,
        "priceBreaks": [
          {
            "price": "0.065",
            "quantity": 100,
          },
          {
            "price": "0.064",
            "quantity": 250,
          },
        ],
        "quantityAvailable": 226274,
        "supplier": "TTI",
        "type": "Bulk",
        "unitPrice": 0.12,
      },
    ],
    "productDoc": "https://www.molex.com/pdm_docs/sd/510210200_sd.pdf",
    "productImageUrl": "https://www.mouser.com/images/molex/images/51021_SPL.jpg",
    "productUrl": "https://www.tti.com/content/ttiinc/en/apps/part-detail.html?partsNumber=0510210200&mfgShortname=MOL&utm=CAV016&channel=link&source=VISUAL%20COMMUNICATIONS%20CO&campaigns=tti-api",
    "sourceParts": [],
    "specifications": "https://www.molex.com/pdm_docs/sd/510210200_sd.pdf",
    "totalStock": 226274,
  },
  {
    "description": "Headers & Wire Housings RECEPTACLE 2P",
    "manufacturerLeadTime": NaN,
    "manufacturerName": "Molex",
    "name": "0510210200",
    "packaging": [
      {
        "minimumOrderQuantity": 42,
        "priceBreaks": [
          {
            "price": "0.065",
            "quantity": 100,
          },
          {
            "price": "0.064",
            "quantity": 250,
          },
        ],
        "quantityAvailable": 226274,
        "supplier": "TTI",
        "type": "Bulk",
        "unitPrice": 0.12,
      },
    ],
    "productDoc": "https://www.molex.com/content/dam/molex/molex-dot-com/products/automated/en-us/salesdrawingpdf/510/51021/510210200_sd.pdf",
    "productImageUrl": "https://www.mouser.com/images/molex/images/51021-0200_SPL.jpg",
    "productUrl": "https://www.tti.com/content/ttiinc/en/apps/part-detail.html?partsNumber=51021%2D0200&mfgShortname=MOL&utm=CAV016&channel=link&source=VISUAL%20COMMUNICATIONS%20CO&campaigns=tti-api",
    "sourceParts": [],
    "specifications": "https://www.molex.com/content/dam/molex/molex-dot-com/products/automated/en-us/salesdrawingpdf/510/51021/510210200_sd.pdf",
    "totalStock": 226274,
  },
]
`;
